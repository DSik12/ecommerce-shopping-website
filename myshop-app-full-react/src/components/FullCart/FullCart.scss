@import "../shared/ScssBreakpoint/Breakpoint.scss";
@mixin btn-update-common($padding-top, $padding-left, $margin-top) {
  color: #fff;
  padding: $padding-top $padding-left;
  margin: $margin-top 15px;
  -webkit-transition: background 0.5s;
  transition: background 0.5s;
  border: 2px solid black;
  border-radius: 4px;
  -webkit-box-shadow: 1px 2px rgba(0, 0, 0, 0.774);
  box-shadow: 1px 2px rgba(0, 0, 0, 0.774);
  font-size: 12px;
  text-decoration: none;
  background-image: -webkit-gradient(
    linear,
    left top,
    left bottom,
    from(#009ce5),
    to(#0096db)
  );
  background-image: linear-gradient(to bottom, #009ce5, #0096db);
  font-weight: bolder;
}
.btn {
  display: inline-block;
  @include btn-update-common(
    $padding-top: 5px,
    $padding-left: 20px,
    $margin-top: 30px
    
  );
}

.update {
  width: 100px;
  @include btn-update-common(
    $padding-top: 8px,
    $padding-left: 25px,
    $margin-top: 15px
  );
}

.check {
  width: 100px;
  @include btn-update-common(
    $padding-top: 8px,
    $padding-left: 15px,
    $margin-top: 15px
  );
}

.full-remove-button {
  background: red;
  color: white;
}

.price {
  font-size: x-small;
}

.full-cart-table {
  font-size: x-small;
  margin-top: 50px;
  width: 100%;
  border-collapse: collapse;
}

.full-cart-name {
  font-size: x-small;
}

.full-cart-product .amount {
  padding-right: 30px;
}

.cross-icon {
  width: 15px;
  height: 15px;
}

.full-cart-th .total {
  padding-right: 90px;
}

input {
  width: 40%;
}

.shopping {
  text-align: left;
  font-size: large;
  font-weight: bold;
  margin: 50px 0 10px 0;
}

.full-cart-sub {
  font-weight: 600;
}

.full-cart-subtotal {
  text-align: right;
  font-weight: bolder;
}

.full-cart-one {
  border: 2px solid black;
  padding: 1px 6px;
}

.full-cart-table .full-cart-th th {
  text-align: left;
  padding: 5px;
  color: #000;
  font-weight: bold;
}

.full-cart-table td {
  padding: 10px 5px;
}

.full-cart-product a {
  color: #5b639c;
  font-size: 12px;
}

#mp1 {
  width: 50px;
  height: 50px;
  margin-right: 10px;
}

.total-price {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.total-price table {
  border-top: 3px solid;
  width: 100%;
  max-width: 350px;
}

.full-cart-product td:last-child {
  text-align: left;
}

.full-cart-table th:last-child {
  text-align: left;
}

table img {
  width: 150px;
  height: 150px;
}

/* Small and medium devices (portrait tablets,large phones[600px] and landscape tablets[768px] and up) */
@media only screen and (min-width: $tablet-portrait) {
  .btn {
    display: inline-block;
    @include btn-update-common(
      $padding-top: 8px,
      $padding-left: 25px,
      $margin-top: 30px
    );
  }
  .update {
    width: 100px;
    @include btn-update-common(
      $padding-top: 8px,
      $padding-left: 15px,
      $margin-top: 15px
    );
  }
  .check {
    width: 100px;
    @include btn-update-common(
      $padding-top: 8px,
      $padding-left: 15px,
      $margin-top: 15px
    );
  }
  .price {
    font-size: large;
  }
  .full-remove-button {
    background: red;
    color: white;
  }
  .full-cart-table {
    font-size: large;
    margin-top: 50px;
    width: 100%;
    border-collapse: collapse;
  }
  .full-cart-name {
    font-size: large;
  }
  .full-cart-product .amount {
    padding-right: 30px;
  }
  .cross-icon {
    width: 15px;
    height: 15px;
  }
  .full-cart-th .total {
    padding-right: 90px;
  }
  .shopping {
    text-align: left;
    font-size: large;
    font-weight: bold;
  }
  .full-cart-sub {
    font-weight: 600;
  }
  .full-cart-subtotal {
    text-align: right;
    font-weight: bolder;
  }
  .full-cart-one {
    border: 2px solid black;
    padding: 1px 6px;
  }
  .full-cart-table .full-cart-th th {
    text-align: left;
    padding: 5px;
    color: #000;
    font-weight: bold;
  }
  .full-cart-table td {
    padding: 10px 5px;
  }
  .full-cart-product a {
    color: #5b639c;
    font-size: 12px;
  }
  #mp1 {
    width: 100px;
    height: 100px;
    margin-right: 10px;
  }
  .total-price {
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-pack: end;
    -ms-flex-pack: end;
    justify-content: flex-end;
  }
  .total-price table {
    border-top: 3px solid;
    width: 100%;
    max-width: 350px;
  }
  .full-cart-product td:last-child {
    text-align: left;
  }
  .full-cart-table th:last-child {
    text-align: left;
  }
}

/* Large devices (laptops/desktops, 992px and up) */
@media only screen and (min-width: $laptop) {
  .full-remove-button {
    background: red;
    width: 100%;
    color: white;
    padding: 6px;
  }
  .full-cart-qty-input {
    width: 20%;
    font-size: 18px;
    text-align: center;
  }
  .btn {
    display: inline-block;
    @include btn-update-common(
      $padding-top: 8px,
      $padding-left: 25px,
      $margin-top: 30px
    );
  }
  .update {
    width: 100px;
    @include btn-update-common(
      $padding-top: 8px,
      $padding-left: 15px,
      $margin-top: 15px
    );
  }
  .check {
    width: 100px;
    @include btn-update-common(
      $padding-top: 8px,
      $padding-left: 15px,
      $margin-top: 15px
    );
  }
  .btn:hover {
    background-image: -webkit-gradient(
      linear,
      left top,
      left bottom,
      from(#00a3ef),
      to(#009ce5)
    );
    background-image: linear-gradient(to bottom, #00a3ef, #009ce5);
    cursor: pointer;
  }
  .update:hover {
    background-image: -webkit-gradient(
      linear,
      left top,
      left bottom,
      from(#00a3ef),
      to(#009ce5)
    );
    background-image: linear-gradient(to bottom, #00a3ef, #009ce5);
    cursor: pointer;
  }
  .check:hover {
    background-image: -webkit-gradient(
      linear,
      left top,
      left bottom,
      from(#00a3ef),
      to(#009ce5)
    );
    background-image: linear-gradient(to bottom, #00a3ef, #009ce5);
    cursor: pointer;
  }
  .price {
    font-size: x-large;
  }
  .full-cart-table {
    font-size: x-large;
    margin-top: 50px;
    width: 100%;
    border-collapse: collapse;
  }
  .full-cart-name {
    font-size: x-large;
  }
  .full-cart-product .amount {
    padding-right: 30px;
  }
  .cross-icon {
    width: 15px;
    height: 15px;
  }
  .full-cart-th .total {
    padding-right: 90px;
  }
  .shopping {
    text-align: left;
    font-size: x-large;
    font-weight: bold;
  }
  .full-cart-sub {
    font-weight: 600;
  }
  .full-cart-subtotal {
    text-align: right;
    font-weight: bolder;
  }
  .full-cart-one {
    border: 2px solid black;
    padding: 1px 6px;
  }
  .full-cart-table .full-cart-th th {
    text-align: left;
    padding: 5px;
    color: #000;
    font-weight: bold;
  }
  .full-cart-table td {
    padding: 10px 5px;
  }
  .full-cart-product a {
    color: #5b639c;
    font-size: 12px;
  }
  #mp1 {
    width: 150px;
    height: 150px;
    margin-right: 10px;
  }
  .total-price {
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-pack: end;
    -ms-flex-pack: end;
    justify-content: flex-end;
  }
  .total-price table {
    border-top: 3px solid;
    width: 100%;
    max-width: 350px;
  }
  .full-cart-product td:last-child {
    text-align: left;
  }
  .full-cart-table th:last-child {
    text-align: left;
  }
}