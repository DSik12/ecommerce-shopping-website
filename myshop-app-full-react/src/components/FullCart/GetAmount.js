import React from "react";

const GetAmount = ({ price }) => {
 
  return (
    <div>
      <span id="fullTotalAmount">Total: Rs.{price}</span>
    </div>
  );
};

export default GetAmount;
