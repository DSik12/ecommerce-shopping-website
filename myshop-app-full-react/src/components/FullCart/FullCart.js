import React, { useEffect, useState, useContext } from "react";
import fetchUrls from "../../config/config.json";
import GetAmount from "./GetAmount"; //for total price of products
import "./FullCart.scss";
import { Link } from "react-router-dom";
import TableHead from "../shared/TableHead";
import { CartQuantityContext} from '../shared/ReducerContextUtility/ReducerContextUtility'


const Fullcart = () => {
  const [data, fetchData] = useState([]);
  let [qty, setQty] = useState();

  const countContext = useContext(CartQuantityContext);

  const updateCartBackend = (url, method, postData) => {
    fetch(url, {
      method: method,
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
      },
      body: JSON.stringify(postData), // posting or patching the stringified postData
    })
      .then((result) => {
        return result.json();
      })
      .then((resp) => {
        if (method === "GET") {
          fetchData(resp);
        }
        else {
          getProducts();
        }
      })
      .catch((err) => {
        console.log(err);
      })
      .finally(() => {
        console.log("Update cart fetch hit.");
      });
  };

  useEffect(() => {
    getProducts();
  }, []);

  //fetching the data from api
  const getProducts = () => {
    updateCartBackend(fetchUrls.cartItemsApi, "GET");
  };

  let totalPrice = 0;
  //delete the products
  const deleteProduct = (id) => {
    let postData = { id: id };
    updateCartBackend(`${fetchUrls.cartItemsApi}/${id}`, "DELETE", postData);
  };  

  //update the products
  const updateProduct = (data) => {
    let quantityList = document.querySelectorAll(".full-cart-qty-input");
    console.log(quantityList);
    for (let i = 0; i < quantityList.length; i++) {
      let qtyId = quantityList[i].id.slice(-1);
      console.log(qtyId);
      let qtyValue = quantityList[i].value;
      let postData = {quantity : qtyValue};
      updateCartBackend(`${fetchUrls.cartItemsApi}/${qtyId}`, "PATCH", postData);
    };
  }

  //This function is used to set the quantity of a to be deleted item's quantity to state variable qty
  //so that context API decrement can be called that many times to update cart count in real time.

  // It is also used to set the quantity of a to be updated item to state variable qty, so that it can be 
  // compared with current value of quantity after updating once to update cart count in real time.
  const handleSetQuantity = (event) =>{
    console.log("Inside handleSetQuantity");
    console.log(event.target.id);
    if((event.target.id) === ("fullRemoveBtn-"+event.target.id.split('-')[1])){
      const itemId=event.target.id.split('-')[1];
      const item = data.filter((obj)=>{
        return obj.id === Number(itemId)
      })
      console.log(item);
      setQty(Number(item[0].quantity))
    }else{
    setQty(event.target.value);
    }
    console.log(qty);
  }
  const checkValue = (event) => {
    if(qty>event.target.value){
      countContext.countDispatch('decrement'); 
    }
    else{
      countContext.countDispatch('increment'); 
    }

  }

  const updateCartQuantityOnDelete = () => {
    console.log("Inside updateCartQuantityOnDelete");
    for(let i=1;i<=qty;i++){
      countContext.countDispatch('decrement'); 
    }
  }

  const callTwoFunctionsDelete = (itemId) =>{
    deleteProduct(itemId);
    updateCartQuantityOnDelete();
  }

  return (
    <div className="cart-page">
      <table className="full-cart-table">
        {/* <thead></thead> */}
        <TableHead></TableHead>
        <tbody id="fullTableBody" onClick={getProducts}>
          {data.map((item, index) => {
            // total price of individual product
            const amount = Number(item.price) * Number(item.quantity);
            // total price of all product
            totalPrice = totalPrice + amount;

            return (
              <tr key={index}>
                <td>
                  <img src={item.thumbnail} alt="not load" />
                </td>
                <td>{item.productName}</td>
                <td
                  className="full-cart-price"
                  id={"fullCartProductPrice" + item.id}
                >
                  {"Rs." + item.price}
                </td>
                <td className="full-cart-qty">
                  <input
                    type="number"
                    className="full-cart-qty-input"
                    id={"fullTotalQuantity" + item.id}
                    min={1}
                    defaultValue={item.quantity}
                    onMouseOver={handleSetQuantity}// This function is used to set present quantity of a product to be updated, to state variable qty on mouse over. 
                    onChange={checkValue}// This function is used to call the context API to update the cart quantity in real time, on updating the quantity of an item.
                  />
                </td>

                <td className="full-amount" id={"fullProductAmount" + item.id}>
                  Rs. {amount}
                </td>
                <td>
                  <button
                    className="full-remove-button"
                    id={"fullRemoveBtn-" + item.id}
                    onMouseOver={handleSetQuantity}// This function is used to set the quantity of a product to be deleted, to state variable qty on mouse over it's delete button. 
                    onClick={callTwoFunctionsDelete.bind(this,item.id)}// This is used to call a function which in turn calls 2 functions in it, on click of delete button for a product
                  >
                    X
                  </button>
                </td>
              </tr>
            );
          })}
        </tbody>
        <tbody>
          <tr className="full-cart-subtotal">
            <td colSpan={6} className="full-cart-sub">
              <GetAmount price={totalPrice} />
            </td>
          </tr>
        </tbody>
      </table>
      <div>
        <button
          type="button"
          className="update"
          id="update"
          onClick={updateProduct.bind(this)}
        >
          Update
        </button>
        <Link to="/checkout/addresscheckout" className="btn">
          Checkout
        </Link>
      </div>

      <div className="total-price">
        <Link to="/checkout/addresscheckout" className="btn">
          Checkout
        </Link>
      </div>
    </div>
  );
};

export default Fullcart;