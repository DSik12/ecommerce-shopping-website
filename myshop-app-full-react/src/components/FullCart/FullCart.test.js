import React from "react";
import ReactDOM from "react-dom";
import FullCart from "./FullCart";
import { BrowserRouter } from "react-router-dom";
import { render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";


describe("FullCart Component", () => {
  it("renders the component properly", () => {
    const divEl = document.createElement("div");
    ReactDOM.render(
      <BrowserRouter>
        <FullCart />
      </BrowserRouter>,
      divEl
    );
  });

  it("should have shopping cart as text", () => {
    render(
      <BrowserRouter>
        <FullCart />
      </BrowserRouter>
    );
    const officialTxtEl = screen.getByText("Shopping Cart");
    expect(officialTxtEl).toBeInTheDocument();
  });

  it("should have view all as text", () => {
    render(
      <BrowserRouter>
        <FullCart />
      </BrowserRouter>
    );
    const officialTxtEl = screen.getByText("Total");
    expect(officialTxtEl).toBeInTheDocument();
  });

  it('should have right FullCart comp snapshot', () => {
      const tree =renderer.create(<BrowserRouter>
        <FullCart />
      </BrowserRouter>).toJSON();
      expect(tree).toMatchSnapshot();
});

});
