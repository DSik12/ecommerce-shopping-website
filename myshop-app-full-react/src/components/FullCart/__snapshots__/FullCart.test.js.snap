// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FullCart Component should have right FullCart comp snapshot 1`] = `
<div
  className="cart-page"
>
  <table
    className="full-cart-table"
  >
    <caption
      className="shopping"
    >
      Shopping Cart
    </caption>
    <thead>
      <tr
        className="full-cart-th"
      >
        <th />
        <th>
          Item(s)
        </th>
        <th>
          Price
        </th>
        <th>
          Qty
        </th>
        <th
          className="total"
        >
          Total
        </th>
      </tr>
    </thead>
    <tbody
      id="fullTableBody"
      onClick={[Function]}
    />
    <tbody>
      <tr
        className="full-cart-subtotal"
      >
        <td
          className="full-cart-sub"
          colSpan={6}
        >
          <div>
            <span
              id="fullTotalAmount"
            >
              Total: Rs.
              0
            </span>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
  <div>
    <button
      className="update"
      id="update"
      onClick={[Function]}
      type="button"
    >
      Update
    </button>
    <a
      className="btn"
      href="/checkout/addresscheckout"
      onClick={[Function]}
    >
      Checkout
    </a>
  </div>
  <div
    className="total-price"
  >
    <a
      className="btn"
      href="/checkout/addresscheckout"
      onClick={[Function]}
    >
      Checkout
    </a>
  </div>
</div>
`;
